Fix spi_flash_at45 driver and sample

anangl opened the pull request:
For some reason, use of the DT_NUM_INST macro in this driver was not
replaced with DT_NUM_INST_STATUS_OKAY. In consequence, this driver
could not be compiled successfully. Furthermore, because this now
undefined macro had been used inside UTIL_LISTIFY, gcc was not even
able to report an error, instead the compilation just "froze".
This patch replaces UTIL_LISTIFY(DT_NUM_INST()) with more appropriate
in this context DT_INST_FOREACH_STATUS_OKAY. It also adds an apparently
missing inclusion of sys/byteorder.h.

The second commit updates the whitelist in the spi_flash_at45 sample description,
so that this sample is again built by sanitycheck and such compilation errors like
the described above cannot go unnoticed.

Two further commits are added in order to improve the test case descriptions
of the sample and get rid of the whitelist in favor of DT based filtering.

anangl committed:
drivers/flash/spi_flash_at45: Fix compilation errors

For some reason, use of the DT_NUM_INST macro in this driver was not
replaced with DT_NUM_INST_STATUS_OKAY. In consequence, this driver
could not be compiled successfully. Furthermore, because this now
undefined macro had been used inside UTIL_LISTIFY, gcc was not even
able to report an error, instead the compilation just "froze".
This patch replaces UTIL_LISTIFY(DT_NUM_INST()) with more appropriate
in this context DT_INST_FOREACH_STATUS_OKAY. It also adds an apparently
missing inclusion of sys/byteorder.h.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
samples/drivers/spi_flash_at45: Update nrf9160_pca10090 board name

This sample still used the old name of the nrf9160dk_nrf9160 board,
both in documentation and sample.yaml. Because of the latter, this
sample was not picked up by sanitycheck, what resulted in compilation
errors in the spi_flash_at45 driver not caught by CI.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

PerMac commented:
@anangl You might want to look at #24943 which relates to the usage of `fixture` for tests requiring additional hardware modifications and filtering those. I would also suggest to modify the sample.yaml and omit using of whitelist attribute as it is being deprecated (#26382). If you think that the building of a sample should be tested by CI I would advise you to use two 'test cases' in the sample.yaml. The first one with `build_only: true` attribute and with `integration: XX` where you define for which platform building of the sample will be tested. The second test case would require a `fixture` and will be run on-target with `--device-testing` only when the proper hardware is provided.

PerMac reviewed (changes requested):
You might want to look at #24943 which relates to the usage of fixture for tests requiring additional hardware modifications and filtering those. I would also suggest to modify the sample.yaml and omit using of whitelist attribute as it is being deprecated (#26382). If you think that the building of a sample should be tested by CI I would advise you to use two 'test cases' in the sample.yaml. The first one with build_only: true attribute and with integration: XX where you define for which platform building of the sample will be tested. The second test case would require a fixture and will be run on-target with --device-testing only when the proper hardware is provided.

PerMac commented on samples/drivers/spi_flash_at45/sample.yaml:
```diff
@@ -3,4 +3,4 @@ sample:
   sample.drivers.spi.flash.at45:
     tags: spi flash
-    platform_whitelist: nrf9160_pca10090
+    platform_whitelist: nrf9160dk_nrf9160
```
I think this sample will fail with TIMEOUT when run by sanitycheck even with correct hardware being provided as the sample does not provide a way to verify its correctness (if `build_only: true` is not used). (You can look at #26187 for an example of how this can be solved)

anangl committed:
samples/drivers/spi_flash_at45: Improve test case descriptions

Instead of whitelist, use a filter that checks if a required
"atmel,at45" compatible node is present.
Add a build-only test case to be used by CI and provide a harness
configuration so that execution of the sample can be also verified.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
@PerMac I added a commit with improvements in the test case descriptions, as you suggested.

mbolivar-nordic commented on drivers/flash/spi_flash_at45.c:
```diff
@@ -703,4 +704,4 @@ static const struct flash_driver_api spi_flash_at45_api = {
 		      &spi_flash_at45_api);
 
-UTIL_LISTIFY(DT_NUM_INST(DT_DRV_COMPAT), SPI_FLASH_AT45_INST, _)
+UTIL_LISTIFY(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT), SPI_FLASH_AT45_INST, _)
```
```suggestion
DT_INST_FOREACH_STATUS_OKAY(SPI_FLASH_AT45_INST)
```

along with this change above:

```
#define SPI_FLASH_AT45_INST(idx)					     \
```

anangl replied:
Applied.

mbolivar-nordic reviewed (changes requested):
The use of UTIL_LISTIFY() in this context is no longer necessary (or recommended), so I suggest a small change to use DT_INST_FOREACH_STATUS_OKAY.

https://docs.zephyrproject.org/latest/guides/dts/howtos.html#write-device-drivers-using-devicetree-apis

anangl committed:
boards: nrf9160dk_nrf9160: Add "spi" as supported feature

Plus corresponding configuration file for the spi_loopback test that
will be activated for that board by this added capability.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
